pkg/util/node/node.go

```diff
@@ -52,7 +54,36 @@ func GetHostname(hostnameOverride string) string {
 		}
 		hostname = nodename
 	}
-	return strings.ToLower(strings.TrimSpace(hostname))
+	hostname = strings.ToLower(strings.TrimSpace(hostname))
+	if msgs := utilvalidation.IsDNS1123Subdomain(hostname); len(msgs) != 0 {
+		glog.Fatalf("Invalid hostname: %s", strings.Join(msgs, ", "))
+	}
+	return hostname
+}
+
+// GetNodeName returns the node name according to the cloud provider
+// if cloud provider is specified. Otherwise, returns the hostname of the node.
+func GetNodeName(cloud cloudprovider.Interface, hostname string) (types.NodeName, error) {
+	if cloud == nil {
+		return types.NodeName(hostname), nil
+	}
+
+	instances, ok := cloud.Instances()
+	if !ok {
+		return "", fmt.Errorf("failed to get instances from cloud provider")
+	}
+
+	nodeName, err := instances.CurrentNodeName(hostname)
+	if err != nil {
+		return "", fmt.Errorf("error fetching current node name from cloud provider: %v", err)
+	}
+
+	if msgs := utilvalidation.IsDNS1123Subdomain(string(nodeName)); len(msgs) != 0 {
+		return "", fmt.Errorf("invalid nodename: %s", strings.Join(msgs, ", "))
+	}
+
+	glog.V(2).Infof("cloud provider determined current node name to be %s", nodeName)
```

Comment:
The first letter of log info should be upper-case. (the original code is lower-case though)